### PR TITLE
Add optional index mapping dir in mmap text datasets

### PR DESCRIPTION
**Pull request by gheinrich**
If datasets are stored on a read-only medium, index files cannot be created into adjacent files and an
alternative directory must be specified for index
mapping files.

# What does this PR do ?

This commit adds an optional flag to specify the directory to write index files to.

**Collection**: nlp

# Changelog 
- Add an optional `index_mapping_dir` to the constructors of `TextMemMapDataset`,  `CSVMemMapDataset` and `JSONLMemMapDataset`.
- Add unit tests.
- Propagate `index_mapping_dir` from `GPTSFTDataset`.
- 
# Usage

```bash
nvidia-docker run --shm-size=4gb  -v ~/llm/data:/llm/data:ro -it sft:latest  python megatron_gpt_sft.py ++model.data.train_ds.index_mapping_dir=/tmp  ++model.data.validation_ds.index_mapping_dir=/tmp
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py [hunk 1]
```diff
@@ -31,11 +33,49 @@
...
+    files. This base name is the base on top of which suffixes
+    like .npy or .info are added.
+
+    Args:
+        fn: filename to get base name for.
+        index_mapping_dir: directory to save the index mapping to.
+                If None, will write to the same folder as the dataset.
+    """
+    if index_mapping_dir:
+        idx_fn = f"{os.path.join(index_mapping_dir, _create_short_hash(fn))}.{__idx_suffix__}"
```
Comment (michalivne) [inline nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py:68]: How about replacing `_create_short_hash(fn)` with `fn`?
Both are relying on the name of the file, but the latter is more transparent (can easily remove only that cache).
If you do so please also call `os.path.makedirs(os.path.join(index_mapping_dir, os.path.dirname(fn)))` to make sure all required paths are created.

File: nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py [hunk 2]
```diff
@@ -31,11 +33,49 @@
...
+    Args:
+        string: the string to create a hash for.
+    Returns:
+        A short hash of the string.
+    """
+    # Compute the SHA-256 hash of the string
+    sha256_hash = hashlib.sha256(string.encode()).hexdigest()
+
+    # Take a portion of the hash to create a short hash
+    short_hash = sha256_hash[:8]
```
Comment (michalivne) [inline nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py:50]: Why use short hash? Why take the risk of collision?

**Review by michalivne — APPROVED** [review]
Great PR! See one comment/suggestion.

File: nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py (hunk 1 above)
Comment (michalivne) [reply nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py:68]: Perfect. I would choose transparency over short names in this case (easier debugging and removal).

Comment (gheinrich) [reply nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py]: Hi, I reworked the patch as per your suggestion. I also added test cases for `CSVMemMapDataset`.

File: nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py (hunk 2 above)
Comment (gheinrich) [reply nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py]: As discussed, I removed this part.

**Comment by gheinrich** [comment]
Hello @michalivne, could you review the latest version of this change? Thank you very much in advance!

File: nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py [hunk 3]
```diff
@@ -246,30 +264,47 @@ class CSVMemMapDataset(TextMemMapDataset):
...
             sort_dataset_paths=sort_dataset_paths,
+            index_mapping_dir=index_mapping_dir,
         )
         self._data_col = data_col
         self._data_sep = data_sep
 
     def _build_data_from_text(self, text):
         """Return a CSV field from text"""
         # get CSV field
+        print("text", text)
```
Comment (michalivne) [inline nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py:305]: Do not forget to remove all `print` debug statements.

File: nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py [hunk 4]
```diff
@@ -304,9 +358,48 @@ def _index_file_exists(idx_fn):
         return False
 
 
-def _build_memmap_index_files(newline_int, build_index_fn, fn):
+def _index_fn(fn: str, index_mapping_dir: str) -> str:
```
Comment (michalivne) [inline nemo/collections/nlp/data/language_modeling/text_memmap_dataset.py:361]: Very useful!

**Review by michalivne — COMMENTED** [review]
LGTM! Very useful PR. Please remove all debug print statements.

**Pull request merged**